bob-common/src/mapper.rs

```diff
@@ -130,19 +171,31 @@ impl Virtual {
         let target_nodes = self.get_target_nodes_for_key(key);
         trace!("extract indexes of target nodes");
         trace!("nodes available: {}", self.nodes.len());
-        let offset = self.support_nodes_offset.fetch_add(1, Ordering::Relaxed);
-        let mut support_nodes = Vec::with_capacity(count);
-        for (id, node) in self.nodes.iter().skip(offset % self.nodes.len()) {
-            if target_nodes.iter().all(|i| i.index() != *id) {
+        let mut support_nodes: Vec<&Node> = Vec::with_capacity(count);
+
+        let starting_index = Self::find_support_node_offset(
+            &self.nodes,
+            target_nodes,
+            self.support_nodes_offset.fetch_add(1, Ordering::Relaxed),
+        );
+
+        let len = self.nodes.len();
+        for i in 0..len {
+            let node = &self.nodes[(i + starting_index) % len];
+            if node.connection_available() && target_nodes.iter().all(|n| n.index() != node.index())
+            {
                 support_nodes.push(node);
                 if support_nodes.len() >= count {
                     break;
                 }
             }
         }
-        if support_nodes.len() < count && offset % self.nodes.len() > 0 {
-            for (id, node) in self.nodes.iter().take(offset % self.nodes.len()) {
-                if target_nodes.iter().all(|i| i.index() != *id) {
+        if support_nodes.len() < count {
```

Comment:
We may not fill the list also if `nodes.len() - target_nodes.len() < count`. So I suggest to extend that condition
```suggestion
        if support_nodes.len() < count && support_nodes.len() +  target_nodes.len() < nodes.len() {
```